### ISSUE TITLE
404 Not Found on OPTIONS request

### ISSUE THREAD

**Issue by 58bits** [fastify/fastify-cors#20]
Completely stumped. I have the cors plugin installed, and two routes as follows:

```js
module.exports = async (app, opts) => {
  app.get('/bar', async (request, reply) => {
    return { hello: 'bar called' }
  })

  app.get('/foo', async (request, reply) => {
    return { hello: 'foo called' }
  })
}
```

OPTIONS requests seem to be working, and are handled by the onRequest hook in the fastify-cors plugin, but when I change the path of the bar route, to anything else that doesn't start with bar... I get a 404 not found response.

e.g.:

```js
module.exports = async (app, opts) => {
  app.get('/fuzzy', async (request, reply) => {
    return { hello: 'bar called' }
  })
}
```

`curl -X OPTIONS http://127.0.0.1:3000/fuzzy`

returns

`{"statusCode":404,"error":"Not Found","message":"Not Found"}`

Any ideas?

**Comment by 58bits** [comment]
I think I've found the problem. Implementing my own wildcard route for OPTIONS requests using `/*` instead of `*` works.
e.g.:
```js
.route({
      method: 'OPTIONS',
      url: '/*',
      handler: async (request, reply) => {
        reply.code(204)
          .header('Content-Length', '0')
          .header('Access-Control-Allow-Origin', '*')
          .header('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
          .send()
      }
    })
```
Is this possibly related to https://github.com/fastify/fastify/issues/81 ?

**Comment by mcollina** [comment]
Can you please add a full repro for your problem? Something I can execute straight away.

It looks like it might be a bug somewhere cc @delvedor

**Comment by 58bits** [comment]
Sure - I'll setup a public repo in a little while.

**Comment by 58bits** [comment]
Try this... https://github.com/58bits/fastify-20

**Comment by mcollina** [comment]
@delvedor can you take a look? This seems something that goes deep down in find-my-way.

**Comment by ljieyao** [comment]
I have the same issue. How to solve this?

**Comment by ljieyao** [comment]
> I think I've found the problem. Implementing my own wildcard route for OPTIONS requests using `/*` instead of `*` works.
> e.g.:
> 
> ```js
> .route({
>       method: 'OPTIONS',
>       url: '/*',
>       handler: async (request, reply) => {
>         reply.code(204)
>           .header('Content-Length', '0')
>           .header('Access-Control-Allow-Origin', '*')
>           .header('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
>           .send()
>       }
>     })
> ```
> 
> Is this possibly related to [fastify/fastify#81](https://github.com/fastify/fastify/issues/81) ?
> 
> I'm using fastify v2.0.1 and fastify-cors v2.1.2 on node v10.15.3

I have tried your workaround. But not work for me

**Comment by 58bits** [comment]
Quick note with more to come - but it may be that the way we are registering app.js - without using fastify-plugin, is affecting the scope of fastify-cors.

**Comment by mcollina** [comment]
it seems you folks a have a better understanding of your issue than us. Would you like to send a PR?

**Comment by ljieyao** [comment]
Not really understand what to do, what do you mean by "the way we are registering app.js"? @58bits

**Comment by 58bits** [comment]
Sorry both @daveamayombo and I are on the road until later this week. Stay tuned...

**Comment by 58bits** [comment]
Apologies for the slow reply. What we _think_ might be happening, is that in our application, we are registering a main app.js plugin, after index.js. https://github.com/58bits/fastify-20/blob/master/index.js - and that if instead the fastify-cors module is loaded in index.js (top level scope?) it works.

We still don't know 100% why, and have not had time to dig into find-my-way, or the problem in more detail.

For now - we've simply created our own plugin, using the `'/*'` catch-all route as follows (leaning heavily on the fastify-cors module as an example).

```js
'use strict'

const fp = require('fastify-plugin')

async function setup (app) {
  // Wildcard OPTIONS handler for CORS preflight requests
  app.route({
    method: 'OPTIONS',
    url: '/*',
    handler: async (request, reply) => {

      var reqAllowedHeaders = request.headers['access-control-request-headers']
      if (reqAllowedHeaders !== undefined) {
        reply.header('Access-Control-Allow-Headers', reqAllowedHeaders)
      }
      reply.code(204)
        .header('Content-Length', '0')
        .header('Access-Control-Allow-Origin', 'http://localhost:8080')
        .header('Access-Control-Allow-Credentials', true)
        .header('Access-Control-Allow-Methods', 'GET,HEAD,PUT,PATCH,POST,DELETE')
        .send()
    }
  })

  // CORS reply - 'Access-Control-Allow-Origin', '*' for now..
  // See https://github.com/fastify/fastify-cors/issues/20
  app.addHook('onRequest', function (request, reply, next) {
    reply.header('Access-Control-Allow-Origin', 'http://localhost:8080')
    reply.header('Access-Control-Allow-Credentials', true)
    next()
  })
}

module.exports = fp(setup)
```

**Comment by alexey-sh** [comment]
@58bits thank you, your example works excellent.

**Comment by StarpTech** [comment]
@58bits please upgrade to the latest version. I think it was fixed with https://github.com/delvedor/find-my-way/pull/133

**Comment by acro5piano** [comment]
**This is a different problem as OP says OPTIONS**, but I struggled to fix `curl -I http://127.0.0.1:3000` returns `404`.

For others who miss the document, this helped me a lot.

```ts
const app = fastify({
    exposeHeadRoutes: true,       // <-- this option
})
```

Related

- https://github.com/fastify/fastify/pull/2700
- https://github.com/fastify/fastify/pull/2826
